File: trading_app.py
import os
from datetime import date
from typing import Dict, Any
from supabase import create_client, Client


class TradingApp:
    def __init__(self, url: str, key: str, starting_cash: float = 1_000_000):
        self.supabase: Client = create_client(url, key)
        self.cash: float = starting_cash
        self.holdings: Dict[str, Dict[str, float]] = {}

    def load_holdings(self):
        """Load existing holdings from database"""
        response = self.supabase.table("holdings").select("symbol, quantity, avg_price").execute()
        
        self.holdings = {}
        for row in response.data:
            self.holdings[row["symbol"]] = {
                "quantity": row["quantity"],
                "avg_price": row["avg_price"]
            }

    def _save_holding(self, symbol: str):
        """Save or delete a holding in the database"""
        if symbol not in self.holdings:
            # Delete the holding if it no longer exists
            self.supabase.table("holdings").delete().eq("symbol", symbol).execute()
        else:
            # Upsert the current holding
            position = self.holdings[symbol]
            self.supabase.table("holdings").upsert({
                "symbol": symbol,
                "quantity": position["quantity"],
                "avg_price": position["avg_price"]
            }, on_conflict="symbol").execute()

    def _get_latest_price(self, symbol: str) -> float:
        response = (
            self.supabase.table("daily_prices")
            .select("close_price")
            .eq("symbol", symbol)
            .order("date", desc=True)
            .limit(1)
            .execute()
        )
        if not response.data:
            raise ValueError(f"No price data for symbol {symbol}")
        return float(response.data[0]["close_price"])

    def buy(self, symbol: str, quantity: int) -> Dict[str, Any]:
        if symbol not in self.holdings and len(self.holdings) >= 10:
            raise ValueError("Maximum number of holdings reached")
        
        price = self._get_latest_price(symbol)
        cost = price * quantity
        if cost > self.cash:
            raise ValueError("Insufficient cash")
        
        self.cash -= cost
        position = self.holdings.get(symbol, {"quantity": 0, "avg_price": 0.0})
        new_quantity = position["quantity"] + quantity
        new_avg_price = (
            position["quantity"] * position["avg_price"] + cost
        ) / new_quantity
        
        self.holdings[symbol] = {"quantity": new_quantity, "avg_price": new_avg_price}
        
        # Save the updated holding to database
        self._save_holding(symbol)
        
        return self.update_daily_pnl()

    def sell(self, symbol: str, quantity: int) -> Dict[str, Any]:
        if symbol not in self.holdings or self.holdings[symbol]["quantity"] < quantity:
            raise ValueError("Not enough shares to sell")
        
        price = self._get_latest_price(symbol)
        proceeds = price * quantity
        self.cash += proceeds
        
        position = self.holdings[symbol]
        position["quantity"] -= quantity
        
        if position["quantity"] == 0:
            del self.holdings[symbol]
        else:
            self.holdings[symbol] = position
        
        # Save the updated (or deleted) holding to database
        self._save_holding(symbol)
        
        return self.update_daily_pnl()

    def _compute_unrealized_pnl(self) -> float:
        pnl = 0.0
        for symbol, position in self.holdings.items():
            price = self._get_latest_price(symbol)
            pnl += (price - position["avg_price"]) * position["quantity"]
        return pnl

    def _portfolio_summary(self) -> Dict[str, Any]:
        """Return cash, holdings, and unrealized P&L totals."""
        pnl = self._compute_unrealized_pnl()
        total = self.cash + pnl
        return {
            "cash": self.cash,
            "holdings": self.holdings,
            "unrealized_pnl": pnl,
            "total": total,
        }

    def update_daily_pnl(self) -> Dict[str, Any]:
        summary = self._portfolio_summary()
        self.supabase.table("pnl").upsert(
            {
                "date": date.today().isoformat(),
                "cash": summary["cash"],
                "unrealized_pnl": summary["unrealized_pnl"],
                "total": summary["total"],
            },
            on_conflict="date",
        ).execute()
        return summary


def create_app() -> TradingApp:
    url = os.environ.get("SUPABASE_URL")
    key = os.environ.get("SUPABASE_KEY")
    if not url or not key:
        raise EnvironmentError("SUPABASE_URL and SUPABASE_KEY must be set")
    
    app = TradingApp(url, key)
    app.load_holdings()  # Load existing holdings from database
    return app


if __name__ == "__main__":
    app = create_app()
    # Example usage
    # app.buy("AAPL", 10)
    # app.sell("AAPL", 5)
    print(app.update_daily_pnl())

File: lib/tradingApp.js
import { createClient } from '@supabase/supabase-js'

export class TradingApp {
  constructor(url, key, startingCash = 1_000_000) {
    this.supabase = createClient(url, key)
    this.cash = startingCash
    this.holdings = {} // symbol -> { quantity, avg_price }
    console.log(`🏁 TradingApp initialized with starting cash: $${startingCash}`)
  }

  async loadHoldings() {
    console.log('📥 Loading holdings from database...')
    
    // Load holdings from database
    const { data: holdingsData, error: holdingsError } = await this.supabase
      .from('holdings')
      .select('symbol, quantity, avg_price')

    if (holdingsError) {
      console.error('❌ Error loading holdings:', holdingsError)
      throw new Error(`Failed to load holdings: ${holdingsError.message}`)
    }

    this.holdings = {}
    for (const row of holdingsData) {
      this.holdings[row.symbol] = {
        quantity: row.quantity,
        avg_price: row.avg_price,
      }
    }
    console.log(`📊 Loaded ${Object.keys(this.holdings).length} holdings:`, this.holdings)

    // Load cash from the most recent P&L record
    console.log('💰 Loading cash from database...')
    const { data: pnlData, error: pnlError } = await this.supabase
      .from('pnl')
      .select('cash')
      .order('date', { ascending: false })
      .limit(1)

    if (pnlError) {
      console.warn('⚠️ Error loading P&L data:', pnlError.message)
      console.log(`💵 Using starting cash: $${this.cash}`)
    } else if (pnlData && pnlData.length > 0) {
      this.cash = pnlData[0].cash
      console.log(`💵 Loaded cash from database: $${this.cash}`)
    } else {
      console.log(`💵 No P&L data found, using starting cash: $${this.cash}`)
    }
  }

  async saveHolding(symbol) {
    console.log(`💾 Saving holding for ${symbol}...`)
    const pos = this.holdings[symbol]
    if (!pos) {
      const { error } = await this.supabase.from('holdings').delete().eq('symbol', symbol)
      if (error) console.error('❌ Error deleting holding:', error)
      return
    }
    
    const { error } = await this.supabase.from('holdings').upsert(
      { symbol, quantity: pos.quantity, avg_price: pos.avg_price },
      { onConflict: 'symbol' }
    )
    
    if (error) {
      console.error('❌ Error saving holding:', error)
      throw new Error(`Failed to save holding: ${error.message}`)
    }
    console.log(`✅ Saved holding: ${symbol}`)
  }

  async getLatestPrice(symbol) {
    console.log(`💰 Getting latest price for ${symbol}...`)
    const { data, error } = await this.supabase
      .from('daily_prices')
      .select('close_price')
      .eq('symbol', symbol)
      .order('date', { ascending: false })
      .limit(1)
      .single()

    if (error) {
      console.error(`❌ Error getting price for ${symbol}:`, error)
      throw new Error(`No price data for ${symbol}: ${error.message}`)
    }
    if (!data) {
      console.error(`❌ No price data found for ${symbol}`)
      throw new Error(`No price data for ${symbol}`)
    }
    
    const price = parseFloat(data.close_price)
    console.log(`📈 Latest price for ${symbol}: $${price}`)
    return price
  }

  async buy(symbol, quantity) {
    console.log(`\n🛒 BUY ORDER: ${quantity} shares of ${symbol}`)
    console.log(`💰 Current cash: $${this.cash}`)
    console.log(`📊 Current holdings:`, this.holdings)

    if (!this.holdings[symbol] && Object.keys(this.holdings).length >= 10) {
      throw new Error("Maximum number of holdings reached")
    }
    
    const price = await this.getLatestPrice(symbol)
    const cost = price * quantity
    
    console.log(`💵 Cost: ${quantity} × $${price} = $${cost}`)
    
    if (cost > this.cash) {
      throw new Error(`Insufficient cash. Need $${cost}, have $${this.cash}`)
    }

    this.cash -= cost
    const prev = this.holdings[symbol] || { quantity: 0, avg_price: 0 }
    const newQty = prev.quantity + quantity
    const newAvg = (prev.quantity * prev.avg_price + cost) / newQty

    this.holdings[symbol] = { quantity: newQty, avg_price: newAvg }
    
    console.log(`💰 New cash: $${this.cash}`)
    console.log(`📊 New position:`, this.holdings[symbol])
    
    await this.saveHolding(symbol)
    const result = await this.updateDailyPnl()
    console.log(`✅ Buy complete!`)
    return result
  }

  async sell(symbol, quantity) {
    console.log(`\n💸 SELL ORDER: ${quantity} shares of ${symbol}`)
    
    const pos = this.holdings[symbol]
    if (!pos || pos.quantity < quantity) {
      throw new Error(`Not enough shares to sell. Have ${pos?.quantity || 0}, trying to sell ${quantity}`)
    }

    const price = await this.getLatestPrice(symbol)
    const proceeds = price * quantity
    this.cash += proceeds

    const remaining = pos.quantity - quantity
    if (remaining === 0) {
      delete this.holdings[symbol]
    } else {
      this.holdings[symbol] = { quantity: remaining, avg_price: pos.avg_price }
    }

    await this.saveHolding(symbol)
    return this.updateDailyPnl()
  }

  async computeUnrealizedPnl() {
    let pnl = 0
    for (const [symbol, pos] of Object.entries(this.holdings)) {
      const price = await this.getLatestPrice(symbol)
      pnl += (price - pos.avg_price) * pos.quantity
    }
    return pnl
  }

  async portfolioSummary() {
    const pnl = await this.computeUnrealizedPnl()
    return {
      cash: this.cash,
      holdings: this.holdings,
      unrealized_pnl: pnl,
      total: this.cash + pnl,
    }
  }

  async updateDailyPnl() {
    const summary = await this.portfolioSummary()
    const today = new Date().toISOString().split("T")[0]

    console.log('📈 Saving P&L summary:', summary)

    const { error } = await this.supabase.from('pnl').upsert(
      {
        date: today,
        cash: summary.cash,
        unrealized_pnl: summary.unrealized_pnl,
        total: summary.total,
      },
      { onConflict: 'date' }
    )

    if (error) {
      console.error('❌ Error saving P&L:', error)
      throw new Error(`Failed to save P&L: ${error.message}`)
    }

    return summary
  }
}
